actix-http/src/helpers.rs

```diff
@@ -106,65 +71,8 @@ pub fn write_content_length(mut n: usize, bytes: &mut BytesMut) {
 
         bytes.put_slice(&buf);
     } else {
-        bytes.put_slice(b"\r\ncontent-length: ");
-        convert_usize(n, bytes);
-    }
-}
-
-pub(crate) fn convert_usize(mut n: usize, bytes: &mut BytesMut) {
-    let mut curr: isize = 39;
-    let mut buf: [u8; 41] = unsafe { mem::uninitialized() };
-    buf[39] = b'\r';
-    buf[40] = b'\n';
-    let buf_ptr = buf.as_mut_ptr();
-    let lut_ptr = DEC_DIGITS_LUT.as_ptr();
-
-    // eagerly decode 4 characters at a time
-    while n >= 10_000 {
-        let rem = (n % 10_000) as isize;
-        n /= 10_000;
-
-        let d1 = (rem / 100) << 1;
-        let d2 = (rem % 100) << 1;
-        curr -= 4;
-        unsafe {
-            ptr::copy_nonoverlapping(lut_ptr.offset(d1), buf_ptr.offset(curr), 2);
-            ptr::copy_nonoverlapping(lut_ptr.offset(d2), buf_ptr.offset(curr + 2), 2);
-        }
-    }
-
-    // if we reach here numbers are <= 9999, so at most 4 chars long
-    let mut n = n as isize; // possibly reduce 64bit math
-
-    // decode 2 more chars, if > 2 chars
-    if n >= 100 {
-        let d1 = (n % 100) << 1;
-        n /= 100;
-        curr -= 2;
-        unsafe {
-            ptr::copy_nonoverlapping(lut_ptr.offset(d1), buf_ptr.offset(curr), 2);
-        }
-    }
-
-    // decode last 1 or 2 chars
-    if n < 10 {
-        curr -= 1;
-        unsafe {
-            *buf_ptr.offset(curr) = (n as u8) + b'0';
-        }
-    } else {
-        let d1 = n << 1;
-        curr -= 2;
-        unsafe {
-            ptr::copy_nonoverlapping(lut_ptr.offset(d1), buf_ptr.offset(curr), 2);
-        }
-    }
-
-    unsafe {
-        bytes.extend_from_slice(slice::from_raw_parts(
-            buf_ptr.offset(curr),
-            41 - curr as usize,
-        ));
+        use std::fmt::Write;
+        let _ignored = write!(bytes, "\r\ncontent-length: {}\r\n", n);
```

Comment:
this is optimization

Reply:
As you can see in the code in the rust std lib (in src/libcore/fmt/num.rs), the same optimization is performed there as in `convert_usize` (with the exact same DEC_DIGITS_LUT even). The generated code is not in any way slower than the previous version (you can see for yourself with `cargo asm`).

Since DEC_DIGITS_LUT, and several other things were identical, i think it was probably duplicated from stdlib in the past, when the compiler was not as good (probably older LLVM version). But today the assembly generated with the patch is the same or better in some cases.

Reply:
could you confirm that asm is the same

Reply:
Both before and after the change, the output of 
`cargo asm --build-type release --rust --comments -- actix_http::helpers::write_content_length` is identical (on x86_64, with rust stable 1,34.1). You can see it [here](https://pastebin.com/Y3HYHnn3).

Before the change (on master at a17ff492), running `cargo asm --build-type release --rust --comments -- actix_http::helpers::convert_usize` gives an error, since the function symbol for `convert_usize` is not found. This suggests it was inlined or eliminated by the LTO.

If you have any performance benchmarks I should run, we can also verify it that way, just in case `cargo asm` gives innacurate results.

Reply:
Actually, i will also double-check with objdump shortly.

Reply:
So i did a more thorough comparison for `write_content_length`.

I disassembled (using objdump -M intel) the `test_httpserver-...` executable, generated by `cargo test --release`.

This is the function on master: https://pastebin.com/SGSewqib
This is the function that uses `write!`: https://pastebin.com/zYK068mv

In fact, optimized in context the second version is much shorter. The one on the master isn't trimmed, it just has the `ret` in the middle (suggesting the inlined `convert_usize` is "cold" to the optimizer).

The symbol in the stdlib called from the patched version of the function is here: https://pastebin.com/0E0LnfKP ; This looks very very similar to the inlined `convert_usize`.

Lastly, i see that the symbol in stdlib is called from a lot of places, even in the test executable. This suggests it could have better cache reuse than if it were repeated in the last part of `write_content_length`.

Reply:
let's split this PR into two. i'd like to bench `write!()` first